INT-4351: Add DiscardAwareFileListFilter

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4351

The `WatchService` reacts to the events in the file system and keep
track ove the events until we poll them, e.g. via `listEligibleFiles()`
in the `WatchServiceDirectoryScanner`.
On the other hand the `SourcePollingChannelAdapter` calls the mentioned
`listEligibleFiles()` according its polling period.
At this moment the `FileListFilter` is applied to the polled files.
It may happen that `LastModifiedFileListFilter` can't accept too young
files yet and they are lost for the future consideration.

* To allow, for example, to retain young files by the
`LastModifiedFileListFilter` judgment for the future cycles add
`DiscardAwareFileListFilter` with the `DiscardCallback` support.
The `WatchServiceDirectoryScanner` now registers such a callback into
the filter and stores discarded files into the `filesToPoll` queue
for the future poll cycle.

**Cherry-pick to 5.0**

<!--
Thanks for contributing to Spring Integration. Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #).

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/master/CONTRIBUTING.adoc).
-->

artembilan committed:
INT-4351: Add DiscardAwareFileListFilter

JIRA: https://jira.spring.io/browse/INT-4351

The `WatchService` reacts to the events in the file system and keep
track ove the events until we poll them, e.g. via `listEligibleFiles()`
in the `WatchServiceDirectoryScanner`.
On the other hand the `SourcePollingChannelAdapter` calls the mentioned
`listEligibleFiles()` according its polling period.
At this moment the `FileListFilter` is applied to the polled files.
It may happen that `LastModifiedFileListFilter` can't accept too young
files yet and they are lost for the future consideration.

* To allow, for example, to retain young files by the
`LastModifiedFileListFilter` judgment for the future cycles add
`DiscardAwareFileListFilter` with the `DiscardCallback` support.
The `WatchServiceDirectoryScanner` now registers such a callback into
the filter and stores discarded files into the `filesToPoll` queue
for the future poll cycle.

**Cherry-pick to 5.0**

artembilan committed:
Fix compilation warnings

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/filters/DiscardAwareFileListFilter.java:
```diff
@@ -0,0 +1,35 @@
+	void addDiscardCallback(DiscardCallback<F> discardCallback);
+
+	@FunctionalInterface
+	interface DiscardCallback<F> {
```
Do we really need a new interface? Why not just `Consumer<File>` ?

garyrussell commented on spring-integration-file/src/main/java/org/springframework/integration/file/filters/CompositeFileListFilter.java:
```diff
@@ -101,6 +107,17 @@ public void close() throws IOException {
+					.stream()
+					.filter(DiscardAwareFileListFilter.class::isInstance)
+					.map(f -> (DiscardAwareFileListFilter<F>) f)
+					.forEach(f -> f.addDiscardCallback(discardCallback));
```
We must only call the callback once per file - remember the composite filter passes all files to all filters.

Or, change the queue to linked set.

artembilan replied:
Well, as far as `CompositeFileListFilter` calls `filterFiles()` for each its delegate, it really a part of the API that `discardCallback` will be called from them. But pay attention that this `CompositeFileListFilter` doesn't call `discardCallback` - it just populates to the delegates.

But I agree: let's reconsider the `WatchServiceScanner` to use a some collection with uniqueness to avoid a double call for the same file from different filters!

garyrussell reviewed (changes requested):
That's all for now. Otherwise technique looks good.

artembilan committed:
* Replace `DiscardCallback` with the plain `Consumer`
* Ensure uniqueness in the `WatchServiceDirectoryScanner` internal
queue via a `Set` implementation, since discard callback may be called
several times for the same file from the `CompositeFileListFilter`
according to its nature
* Add JavaDocs and Docs

artembilan committed:
Change `@since` to `5.0.5`

garyrussell commented:
Merged as 0ef659b8a7c2833640e5f95246225893f03ffc95

yihfenglow commented:
Would this fix the issue where if you use .preventDuplicate(true) and a LastModifiedFileListFilter it behaves unexpectedly due to https://stackoverflow.com/questions/39604652/acceptoncefilefilter-keeps-other-filters-from-working-in-a-compositefilelistfilt? 

I believes this quirk throws a lot of people off. If this doesn't fix it, it feels like something that can potentially be extended to do so?

Thanks!

artembilan commented:
I think that question is not related to the problem we have fixed here. I believe what you need is called `ChainFileListFilter` instead of `CompositeFileListFilter`: https://docs.spring.io/spring-integration/docs/5.0.4.RELEASE/reference/html/files.html#file-reading

yihfenglow commented:
What I meant was that if you use .preventDuplicates(true) when setting up the file watcher, it defaults into a CompositeFileListFilter, and when used with a LastModifiedFileListFilter it behaves unexpectedly. 

You can work around it by using a ChainFileListFilter instead, but that often throws people off. 

Hope that makes sense.

artembilan commented:
I’m not sure what you mean with the “throws people off”, but GH is not the place to discuss opinions. 

If you have strong position how that should be, feel free to raise a JIRA to discuss. This one was about exactly `LastModified`, but not `AcceptOnce`.
